File: src/bot/bot.py
"""Discord bot implementation for Garment Sustainability Bot."""

import discord
from discord.ext import commands
from typing import Optional

from config.settings import config
from utils.logger import setup_logger, get_logger
from models import init_db

# Set up logger
logger = setup_logger("gsbot", config.LOG_LEVEL, config.LOG_FILE)


class GarmentSustainabilityBot(commands.Bot):
    """Main bot class for Garment Sustainability Bot."""

    def __init__(self):
        """Initialize the bot."""
        intents = discord.Intents.default()
        intents.message_content = True
        intents.members = True

        super().__init__(
            command_prefix=config.DISCORD_PREFIX,
            intents=intents,
            help_command=commands.DefaultHelpCommand(),
        )

        self.logger = get_logger("gsbot")

    async def setup_hook(self):
        """Set up the bot before it starts."""
        self.logger.info("Setting up Garment Sustainability Bot...")

        # Initialize database
        try:
            init_db()
            self.logger.info("Database initialized successfully")
        except Exception as e:
            self.logger.error(f"Failed to initialize database: {e}")
            raise

        # Load cogs (command groups)
        await self.load_cogs()

        self.logger.info("Bot setup complete")

    async def load_cogs(self):
        """Load all cog modules."""
        cogs = [
            "bot.cogs.sustainability",
            "bot.cogs.materials",
            "bot.cogs.brands",
            "bot.cogs.user_commands",
            "bot.cogs.admin",
        ]

        for cog in cogs:
            try:
                await self.load_extension(cog)
                self.logger.info(f"Loaded cog: {cog}")
            except Exception as e:
                self.logger.error(f"Failed to load cog {cog}: {e}")

    async def on_ready(self):
        """Called when the bot is ready."""
        self.logger.info(f"Logged in as {self.user.name} ({self.user.id})")
        self.logger.info(f"Connected to {len(self.guilds)} guilds")

        # Set bot status
        await self.change_presence(
            activity=discord.Activity(
                type=discord.ActivityType.watching,
                name=f"sustainable fashion | {config.DISCORD_PREFIX}help"
            )
        )

    async def on_command_error(self, ctx: commands.Context, error: Exception):
        """Handle command errors."""
        if isinstance(error, commands.CommandNotFound):
            await ctx.send(
                f"❌ Command not found. Use `{config.DISCORD_PREFIX}help` to see available commands."
            )
        elif isinstance(error, commands.MissingRequiredArgument):
            await ctx.send(f"❌ Missing required argument: {error.param.name}")
        elif isinstance(error, commands.MissingPermissions):
            await ctx.send("❌ You don't have permission to use this command.")
        else:
            self.logger.error(f"Command error: {error}", exc_info=True)
            await ctx.send("❌ An error occurred while processing the command.")

    async def on_message(self, message: discord.Message):
        """Process messages."""
        # Ignore messages from the bot itself
        if message.author == self.user:
            return

        # Process commands
        await self.process_commands(message)


def create_bot() -> GarmentSustainabilityBot:
    """
    Create and return a bot instance.

    Returns:
        GarmentSustainabilityBot instance
    """
    return GarmentSustainabilityBot()

File: src/bot.rs
// SPDX-License-Identifier: MPL-2.0
//! Bot wiring — poise/serenity equivalent of `bot/bot.py`: prefix, intents,
//! presence, cog loading, and the command-error message mapping.

use anyhow::Result;
use poise::serenity_prelude as serenity;

use gsbot::{commands, config::Config, Data, Error};

/// Map framework errors to the same user-facing messages as the Python
/// `on_command_error` handler.
async fn on_error(error: poise::FrameworkError<'_, Data, Error>) {
    use poise::FrameworkError as FE;
    match error {
        FE::UnknownCommand {
            ctx, msg, prefix, ..
        } => {
            let _ = msg
                .channel_id
                .say(
                    &ctx.http,
                    format!(
                        "❌ Command not found. Use `{prefix}help` to see available commands."
                    ),
                )
                .await;
        }
        FE::ArgumentParse { ctx, input, .. } => {
            let what = input.unwrap_or_default();
            let _ = ctx
                .say(format!("❌ Missing required argument: {what}"))
                .await;
        }
        FE::CommandCheckFailed { ctx, .. } => {
            let _ = ctx
                .say("❌ You don't have permission to use this command.")
                .await;
        }
        FE::Command { error, ctx, .. } => {
            tracing::error!("Command error: {error:?}");
            let _ = ctx
                .say("❌ An error occurred while processing the command.")
                .await;
        }
        other => {
            if let Err(e) = poise::builtins::on_error(other).await {
                tracing::error!("Error in error handler: {e:?}");
            }
        }
    }
}

/// Build and run the bot. Equivalent to `create_bot()` + `bot.start()`.
pub async fn run(config: Config, db: sqlx::SqlitePool) -> Result<()> {
    let token = config.discord_token.clone();
    let prefix = config.discord_prefix.clone();
    let presence = format!("sustainable fashion | {prefix}help");

    let options = poise::FrameworkOptions {
        commands: commands::all(),
        prefix_options: poise::PrefixFrameworkOptions {
            prefix: Some(prefix),
            ..Default::default()
        },
        on_error: |e| Box::pin(on_error(e)),
        ..Default::default()
    };

    let framework = poise::Framework::builder()
        .options(options)
        .setup(move |ctx, ready, _framework| {
            Box::pin(async move {
                tracing::info!(
                    "Logged in as {} ({})",
                    ready.user.name,
                    ready.user.id
                );
                ctx.set_activity(Some(serenity::ActivityData::watching(presence)));
                Ok(Data { db, config })
            })
        })
        .build();

    let intents = serenity::GatewayIntents::GUILDS
        | serenity::GatewayIntents::GUILD_MESSAGES
        | serenity::GatewayIntents::MESSAGE_CONTENT
        | serenity::GatewayIntents::GUILD_MEMBERS;

    let mut client = serenity::ClientBuilder::new(token, intents)
        .framework(framework)
        .await?;

    tracing::info!("Starting Garment Sustainability Bot...");
    client.start().await?;
    Ok(())
}
